reactor-extra/src/main/java/reactor/retry/ReconnectMono.java

```diff
@@ -0,0 +1,391 @@
+
+	@SuppressWarnings("unchecked")
+	public static <T> ReconnectMono<T> create(Mono<T> source, Consumer<? super T> onValueExpired) {
+		return create(source, onValueExpired, NOOP_ON_VALUE_RECEIVED);
+	}
+
+	public static <T> ReconnectMono<T> create(Mono<T> source,
+			Consumer<? super T> onValueExpired,
+			BiConsumer<? super T, Invalidate> onValueReceived) {
+		return new ReconnectMono<>(source, onValueExpired, onValueReceived);
+	}
+
+
+	final Mono<T>                            source;
+	final BiConsumer<? super T, Invalidate>  onValueReceived;
+	final Consumer<? super T>                onValueExpired;
+	final ReconnectMainSubscriber<? super T> mainSubscriber;
+
+	volatile int wip;
+	@SuppressWarnings("rawtypes")
+	static final AtomicIntegerFieldUpdater<ReconnectMono> WIP =
+		AtomicIntegerFieldUpdater.newUpdater(ReconnectMono.class, "wip");
+
+	volatile ReconnectInner<T>[] subscribers;
+	@SuppressWarnings("rawtypes")
+	static final AtomicReferenceFieldUpdater<ReconnectMono, ReconnectInner[]> SUBSCRIBERS =
+			AtomicReferenceFieldUpdater.newUpdater(ReconnectMono.class, ReconnectInner[].class, "subscribers");
+
+	@SuppressWarnings("rawtypes")
+	static final ReconnectInner[] EMPTY_UNSUBSCRIBED = new ReconnectInner[0];
+	@SuppressWarnings("rawtypes")
+	static final ReconnectInner[] EMPTY_SUBSCRIBED   = new ReconnectInner[0];
+	@SuppressWarnings("rawtypes")
+	static final ReconnectInner[] READY              = new ReconnectInner[0];
+	@SuppressWarnings("rawtypes")
+	static final ReconnectInner[] TERMINATED         = new ReconnectInner[0];
+
+	static final int ADDED_STATE      = 0;
+	static final int READY_STATE      = 1;
+	static final int TERMINATED_STATE = 2;
+
+	T         value;
+	Throwable t;
+
+	ReconnectMono(
+		Mono<T> source,
+		Consumer<? super T> onValueExpired,
+ 		BiConsumer<? super T, Invalidate> onValueReceived
+	) {
+		this.source = source;
+		this.onValueExpired = onValueExpired;
+		this.onValueReceived = onValueReceived;
+		this.mainSubscriber = new ReconnectMainSubscriber<>(this);
+
+		SUBSCRIBERS.lazySet(this, EMPTY_UNSUBSCRIBED);
+	}
+
+	@Override
+	public Object scanUnsafe(Attr key) {
+		if (key == Attr.PARENT) return source;
+		if (key == Attr.PREFETCH) return Integer.MAX_VALUE;
+
+		final boolean isDisposed = isDisposed();
+		if (key == Attr.TERMINATED) return isDisposed;
+		if (key == Attr.ERROR) return t;
+
+		return null;
+	}
+
+	@Override
+	public void dispose() {
+		this.terminate(new CancellationException("ReconnectMono has already been disposed"));
+	}
+
+	@Override
+	public boolean isDisposed() {
+		return this.subscribers == TERMINATED;
+	}
+
+	@Override
+	@SuppressWarnings("uncheked")
+	public void subscribe(CoreSubscriber<? super T> actual) {
+		final ReconnectInner<T> toAdd = new ReconnectInner<>(actual, this);
+		actual.onSubscribe(toAdd);
+
+		final int state = this.add(toAdd);
+
+		if (state == READY_STATE) {
+			actual.onNext(this.value);
+			actual.onComplete();
```

Comment:
these two lines don't account for the possibility that the `actual` `Subscriber` hasn't necessarily immediately requested upon `onSubscribe` (even though this is usually the case for `Mono`s).

this should be `toAdd.complete(this.value);`, as you correctly do in the `complete()` method.

Reply:
Right. Missed that